### PR TITLE
User Documentation Proposal

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR

- [Design docs](https://docs.google.com/document/d/1wvCvM6pLQ17Zr0DcpS9es2vOFQsNeOaNN99ufGw8O1U/edit#heading=h.w5wnz0kbiild)
- Add description entity model
- Add a new attribute `docs` to the task / workflow
- Add a new attribute `docs` to the taskSpec / workflowSpec, so description entity will be registered when we register a task or workflow
 

Two ways to add documentation for the task
1. In decorator 
```python
@task(docs=Documentation(short_description="hello world", long_description=LongDescription(uri="s3://bucket/file")))
def t1(a: int) -> int:
    ...
```

2. In comment
```python
@task
def t1(a: int) -> int:
    """
    Set the temperature value.

    The value of the temp parameter is stored as a value in
    the class variable temperature.
    """
    return a
```

- flytekit will offload the long description to the s3 bucket if size exceeds 4KB
- flytekit will update the source link by looking up the git remote url

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
https://github.com/flyteorg/flyte/issues/531

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by flixr** [comment]
Nice! But what about e.g. a `ContainerTask`?

**Comment by pingsutw** [comment]
> Nice! But what about e.g. a ContainerTask?

We can add description to all kind of the tasks.

File: flytekit/models/documentation.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    formatting.
+    """
+
+    class DescriptionFormat(Enum):
+        UNKNOWN = 0
+        MARKDOWN = 1
+        HTML = 2
+        RST = 3
+
+    value: Optional[str] = ""
```
Comment (kumare3) [inline flytekit/models/documentation.py:24]: none?

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
